nni/retiarii/oneshot/pytorch/utils.py

```diff
@@ -180,3 +181,205 @@ def replace_input_choice(root_module, init_fn, modules=None):
+        super().__init__(None)
+    
+    def __iter__(self):
+        self.train_iter = iter(self.train_loader)
+        self.val_iter = iter(self.val_loader)
+        return self
+    
+    def __next__(self):
+        try:
+            train_batch = next(self.train_iter)
+        except StopIteration:
+            # training data is used up
+            if self.equal_len or self.train_longger:
+                # if training is the longger one or equal, stop iteration
+                raise StopIteration()
+            # if training is the shorter one, upsample it
+            self.train_iter = iter(self.train_loader)
+            train_batch = next(self.train_iter)
+
+        try:
+            val_batch = next(self.val_iter)
+        except StopIteration:
+            # validation data is used up
+            if not self.train_longger:
+                # if validation is the longger one (the equal condition is 
+                # covered above), stop iteration
+                raise StopIteration()
+            # if validation is the shorter one, upsample it
+            self.val_iter = iter(self.val_loader)
+            val_batch = next(self.val_iter)
+        
+        return train_batch, val_batch
+    
+    def __len__(self) -> int:
+        return max(len(self.train_loader), len(self.val_loader))
+
+class ConcatenateTrainValDataLoader(DataLoader):
+    ''' 
+    A dataloader that yields validation data after training data in an epoch. You will
+    got a batch in a form of (batch, b) in the training step, where the bool variable 'b'
+    is set to true if the 'batch' comes from the training dataset, and False for validation
+    dataset. Some NAS algorithms, i.e. ENAS, may require this type of dataset. 
+
+    Warnings
+    ----------    
+    If you got in some trouble when running with this class, i.e. under a multi-card condition,
+    please try ConcatenateTrainValDataset instead.
+
+    Example
+    --------
+    Fit your dataloaders into a concatenated one.
+    >>> concat_loader = ParallelTrainValDataLoader(train_dataloader, val_datalodaer)
+    Then you can use the 'concat_loader' as a normal training loader.
+    '''
+    def __init__(self, train_dataloader, val_dataloader):
+        '''
+        Parameters
+        ----------
+        train_data : DataLoader
+            training dataloader
+        val_data : DataLoader
+            validation dataloader
+        '''
+        self.train_loader = train_dataloader
+        self.val_loader = val_dataloader
+        super().__init__(None)
+    
+    def __iter__(self):
+        self.cur_iter = iter(self.train_loader)
+        self.is_train = True
+        return self
+    
+    def __next__(self):
+        try:
+            batch = next(self.cur_iter)
+        except StopIteration:
+            # training data is used up, change to validation data
+            if self.is_train:
+                print('change to val data')
```

Comment:
Avoid using `print()` as we are building a tool. We don't want to pollute users' screen. Try to use `logger.info(...)` instead. You can search NNI's code directory for example usages.